### PR TITLE
Implement Dummy HASH

### LINKED ISSUE #11847: Dummy fails when is re-evaluated in S and srepr
Hi all!

Well, i found this little problem, Dummy is a nice function but in the way is implemented exist this:
```
>>> alpha = Dummy("alpha")
>>> i = Integral(1/sqrt(1 - sin(alpha)**2), (alpha, 0, pi/2))
>>> N(i)
0.e+2
>>> i = S(srepr(i))
>>> N(i)
Integral(1/sqrt(-sin(_alpha)**2 + 1), (_alpha, 0, pi/2))
```
Basically, if you works with Dummy, and you get the expression with srepr or similar, when you eval it every Dummy will be interpreted as a new Dummy, so it fails, in the example you can see we can't eval the second expression because the 2 Dummy expression are interpreted as differents, other ex:
```
>>> m=Dummy("m")
>>> a=Matrix([m, m])
>>> a[0]==a[1]
True
>>> b=S(srepr(a))
>>> b[0]==b[1]
False
```

So thinking a solution can be use a hash or similar function, like:
```
>>> m=Dummy("m")
>>> srepr(m)
"Dummy('m', hash=987654321)"
>>> m=Dummy("m")
>>> srepr(m)
"Dummy('m', hash=754619474)"
```
Always will exist the hash collision, but at least to cover all possible can be sympy check the existents Dummy expressions and check the hash never by equal to other, the probability of the hash collision, to be equal, in the same context, with the same name, and in the same expression, i think is very low.

Maybe instead use a hash can be a random expression, with chars like ```8waerfn23jb89a```, that can help to avoid hash collision.

Thx. Cya.

### PR DESCRIPTION
**Pull request by latot**
Fixes https://github.com/sympy/sympy/issues/11847

Hi, well, this try fix the issue, sadly i can't found an easy way to this, basically for 2 words...
"HASH Collision", lets think the example only store the dummy_index and pass it to the Dummy function:

```
>>> a=Dummy()  %Index 0
>>> a = srepr(a)  %Index 0
%% Move to other env to can import the data
>>> b = Dummy() %Index 0
>>> a = eval(a)  %Eval the previous expression, Index 0
>>> b == a  %Hash collision, this two should be differents.
True
```

I think my code normally is... ugly,  so if someone wanna change formats, names, etc, welcome, and any recommendation and discuss obvs this is open.

I choose that method for random number only for cryptographic security, and avoid the most possible the hash collision.

Thx. Cya.

### CODE REVIEW

**Comment by asmeurer** [comment]
The hash stuff seems like overkill to me. I would just save and restore the dummy index. I don't think collisions are actually that big of a deal (as you and @cbm755 pointed out in the issue, two dummys that should be equal being unequal is a bigger issue). If anything, we can change the code from `Dummy._count += 1` to `Dummy._count = obj.dummy_index + 1`. That way, whenever a dummy is created with an explicit dummy index (i.e., from a eval'd srepr string), no new dummy created after that will be equal (unless explicitly set so).

**Comment by latot** [comment]
Hi, i'm trying to interpret your intentions, i can imagine 2 ways for it, but sadly no one can cover this, the basic example i can imagine with it:
```
a = srepr(Dummy()) %%Index 0
%% Go to other env
b = Dummy() %% Index 0
c = eval(a) %% Index 1 (0 + 1)
b = Dummy() %% Index = 1 %% Collision
```
if we move Dummy._count to the input value we will end where we start....

I misread something?

Thx. Cya.

**Comment by asmeurer** [comment]
My suggestion is something like 

```diff
diff --git a/sympy/core/symbol.py b/sympy/core/symbol.py
index 75e749f..2b7ce22 100644
--- a/sympy/core/symbol.py
+++ b/sympy/core/symbol.py
@@ -200,14 +200,19 @@ class Dummy(Symbol):
     is_Dummy = True

     def __new__(cls, name=None, **assumptions):
+        _dummy_index = assumptions.pop('_dummy_index', None)
         if name is None:
             name = "Dummy_" + str(Dummy._count)

         cls._sanitize(assumptions, cls)
         obj = Symbol.__xnew__(cls, name, **assumptions)

-        Dummy._count += 1
-        obj.dummy_index = Dummy._count
+        if _dummy_index:
+            obj.dummy_index = _dummy_index
+            Dummy._count = _dummy_index + 1
+        else:
+            Dummy._count += 1
+            obj.dummy_index = Dummy._count
         return obj

     def __getstate__(self):
```

The issue potentially still exists in cases like:

```
b = Dummy() # index 0
d = eval(a) # index 0
c = Dummy() # index 1
```

Here `d`, a dummy from the other environment, would be inadvertently equal to `b`. I suspect this isn't actually a huge issue. If you can think of where it might be a problem let me know. 

A solution for it would be to capture an index as above, but set the actual dummy index as `Dummy._count + _dummy_index`. That way, new dummys are always unequal to existing ones, but recreated ones still have the same equality. Something like

```diff
diff --git a/sympy/core/symbol.py b/sympy/core/symbol.py
index 75e749f..499baa3 100644
--- a/sympy/core/symbol.py
+++ b/sympy/core/symbol.py
@@ -200,14 +200,17 @@ class Dummy(Symbol):
     is_Dummy = True

     def __new__(cls, name=None, **assumptions):
+        # Not necessarily the final dummy_index
+        _dummy_index = assumptions.pop('_dummy_index', 0)
+
         if name is None:
             name = "Dummy_" + str(Dummy._count)

         cls._sanitize(assumptions, cls)
         obj = Symbol.__xnew__(cls, name, **assumptions)

-        Dummy._count += 1
-        obj.dummy_index = Dummy._count
+        obj.dummy_index = Dummy._count + _dummy_index
+        Dummy._count += _dummy_index + 1
         return obj

     def __getstate__(self):
```

**Comment by asmeurer** [comment]
Ah, scratch that last one. It doesn't work right. The only way to avoid collisions is to keep track of the dummy indices on the Dummy class (rather than just a counter). Or use random hashing. Either way seems overkill personally, since I don't see it as a real problem.

**Comment by latot** [comment]
mm, the problem is when apps export data from python, when the data is re-evaluated all dummy symbols breaks the expressions,,,,

I simplify a little the code, check a little, should be more easy to understand,

**Comment by latot** [comment]
Hi!, i think this is ready, at least the algorithm, is very probable some formal things are missing, well first i'll explain a little some parts:

1) Now every dummy can have a unique random HASH
2) This works with a registry where are saved all hashes to check we always create unique Dummys
3) The HASH to be relative unique it uses the number of registered HASH and a random number of 24 length, this is important to avoid the limit of possibles hashes and set a context to avoid collisions.
4) The dummy function now check when is called 'dummy_index' manually, because if is changed manually and the new HASH don't is recorded we will get errors, and it helps to centralize some checks.
5) I update the 'function' file with this, and fix a bug, the HASH never can be negative, in this system and the previous, because with some codes like 'ccode' and similars this symbol '-' breaks the expression.
6) I change some tests for 2 reasons, one of them is because is using the previous system so we don't need it now, the second tests, is because some codes generate expressions with the HASH, but how is long now some times it transform the result with new lines breaking the test (i check this, results are equals).

Notes:
I don't like the idea set ```shash``` from assumptions, because don't is an assumption, is a property (only to get clear concept of it).
Following @cbm755 advice i'll leave the level of changes after here, so i'll fix and do little changes if is necessary but not improve new things, i think that can will go in other pr.
I would like merge ```shash``` and ```dummy_index```, personally i prefer the first name, shash(symbol hash), i think goes according with the concept, and with this we will not use exactly a 'index' method to sort this.

What do you think?

About where this bug affects, i can imagine 2 generalized cases, first when the user use srepr to export their own data and save it for the future, if you do complex, or long calcs is very useful save the expressions, the second case, when apps use this to import/export data continuously, like octsympy and similars.

Thx. Cya.

**Comment by cbm755** [comment]
I've started reviewing this.  I like the general idea but some things could be simpler or easier to understand.  I added some more commits in #12144.  You can pull from my `latot-dummy-hash` branch if you want.

The default length of the "hash" could be discussed: I don't think its too important.  It also might be nice to use a string (as you originally had), perhaps after understanding the chain rule bits...

I think `_count` should continue doing what it did before (have changed that in my branch).

File: sympy/core/function.py [hunk 1]
```diff
@@ -707,8 +707,7 @@ def fdiff(self, argindex=1):
             else:
                 return Derivative(self, self.args[argindex - 1], evaluate=False)
         # See issue 4624 and issue 4719 and issue 5600
-        arg_dummy = Dummy('xi_%i' % argindex)
-        arg_dummy.dummy_index = hash(self.args[argindex - 1])
+        arg_dummy = Dummy('xi_%i' % argindex, shash=abs(hash(self.args[argindex - 1])))
```
Comment (cbm755) [inline sympy/core/function.py]: Do any tests fail if we just do `arg_dummy = Dummy('xi_%i' % argindex)` here?  I haven't read everything in detail but I assume this is `.dummy_index = ` bit is trying to solve a similar problem to the one we're looking at.

File: sympy/core/function.py [hunk 2]
```diff
@@ -1178,8 +1177,7 @@ def __new__(cls, expr, *variables, **assumptions):
                 obj = None
             else:
                 if not is_symbol:
-                    new_v = Dummy('xi_%i' % i)
-                    new_v.dummy_index = hash(v)
+                    new_v = Dummy('xi_%i' % i, shash=abs(hash(v)))
```
Comment (cbm755) [inline sympy/core/function.py]: Or is it that this `new_v` has to match `arg_dummy` above?

File: sympy/core/symbol.py [hunk 3]
```diff
@@ -216,11 +221,17 @@ def __getstate__(self):
     @cacheit
     def sort_key(self, order=None):
         return self.class_key(), (
-            2, (str(self), self.dummy_index)), S.One.sort_key(), S.One
+            2, (str(self), Dummy._count.index(self.dummy_index))), S.One.sort_key(), S.One
```
Comment (cbm755) [inline sympy/core/symbol.py]: I'm slightly worried about performance here.  Probably `.index` is an O(n) search in the number of Dummy variables right?  If someone using a pattern with lots of Dummies in a tight loop, maybe its a problem.

I think I will record the `_count` in a dict keyed by `dummy_index` instead.  Then O(1) lookup here.

File: sympy/core/function.py (hunk 1 above)
Comment (cbm755) [inline sympy/core/function.py]: I didn't find a test, but I think I know a situation that uses this.  I will file another issue for a test.

**Pull request merged**